sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -142,3 +164,415 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
     result
   }
 }
+
+final class GenericIntArrayData(private val primitiveArray: Array[Int])
+  extends GenericArrayData(Array.empty) {
+  override def array(): Array[Any] = primitiveArray.toArray
+
+  override def copy(): ArrayData = new GenericIntArrayData(primitiveArray)
+
+  override def numElements(): Int = primitiveArray.length
+
+  override def isNullAt(ordinal: Int): Boolean = false
+  override def getInt(ordinal: Int): Int = primitiveArray(ordinal)
+  override def toIntArray(): Array[Int] = {
+    val array = new Array[Int](numElements)
+    System.arraycopy(primitiveArray, 0, array, 0, numElements)
+    array
+  }
+  override def toString(): String = primitiveArray.mkString("[", ",", "]")
+
+  override def equals(o: Any): Boolean = {
```

Comment:
I should implement ```equals()``` and ```hashCode()``` for classes related to ```GenericArrayData```

I already implemented type specialized ```equals()``` and ```hashCode()``` in ```Generic<PrimitiveTypeArray>ArrayData```. An issue in ```equals()``` and ```hashCode``` of ```GenericRefArrayData``` is that a type of each element may be different since they are hold in ```Array[Any]```.
 
If I misunderstood your suggestion, could you please let me know?